syncano_cli/hosting/utils.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+import os
+import sys
+import time
+
+from syncano_cli.logger import get_logger
+
+LOG = get_logger('syncano-hosting')
+
+
+class HostingCommands(object):
+
+    def __init__(self, instance):
+        self.instance = instance
+
+    def list_hosting(self):
+        return [
+            (hosting.label, hosting.domains) for hosting in self.instance.hostings.all()
+        ]
+
+    def list_hosting_files(self, domain):
+        hosting = self._get_hosting(domain=domain)
+        if not hosting:
+            LOG.warn('No default hosting found. Exit.')
+            sys.exit(1)
+
+        files_list = hosting.list_files()
+        return files_list
+
+    def publish(self, domain, base_dir):
+        uploaded_files = []
+        hosting = self._get_hosting(domain=domain)
+        if not hosting:
+            # create a new hosting if no default is present;
+            hosting = self.create_hosting(label='Default hosting', domain=domain)
+
+        for folder, subs, files in os.walk(base_dir):
+            path = folder.split(base_dir)[1][1:]  # skip the /
+            for single_file in files:
+                if path:
+                    file_path = '{}/{}'.format(path, single_file)
+                else:
+                    file_path = single_file
+
+                sys_path = os.path.join(folder, single_file)
+                with open(sys_path, 'rb') as upload_file:
+                    LOG.info('Uploading file: {}'.format(file_path))
+                    hosting.upload_file(path=file_path, file=upload_file)
+
+                uploaded_files.append(file_path)
+                time.sleep(1)  # avoid throttling;
```

Comment:
All this throttling thing is a little bit tricky - because client have no information about other requests (made by others clients) - so your idea to implement retry-after is the best way to handle it (servers knows how many requests should be handled - from all clients).

Will change to 0.02 here.